web/api/middleware/authenticate.go

```diff
@@ -25,36 +30,87 @@ func Authenticate(next http.Handler) http.Handler {
+		authenticateOnSpotify := false
+		redisConn := redisPool.Get()
+		defer redisConn.Close()
+		sessionData, err := redis.StringMap(redisConn.Do("HGETALL", getSessionKey(spotifyToken)))
 		if err != nil {
-			common.Fail(w, fmt.Errorf("Could not build Spotify profile request: %s", err), http.StatusInternalServerError)
-			return
-		}
-		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", spotifyToken))
-		res, err := common.SpotifyClient.Do(req)
-		if err != nil {
-			common.Fail(w, fmt.Errorf("Could not make Spotify profile request: %s", err), http.StatusInternalServerError)
-			return
-		}
-		defer res.Body.Close()
-		if res.StatusCode != http.StatusOK {
-			common.Fail(w, fmt.Errorf("Spotify profile request responded with %d", res.StatusCode), http.StatusForbidden)
-			return
-		}
-		body, err := ioutil.ReadAll(res.Body)
-		if err != nil {
-			common.Fail(w, fmt.Errorf("Could not read Spotify profile response: %s", err), http.StatusInternalServerError)
-			return
+			if err == redis.ErrNil {
+				authenticateOnSpotify = true
+			} else {
+				common.Fail(w, fmt.Errorf("Could not access Redis: %s", err), http.StatusInternalServerError)
+				return
+			}
+		} else if sessionData["id"] == "" || sessionData["country"] == "" {
+			// We have to have country and ID for the user, name is not as important
+			authenticateOnSpotify = true
 		}
 		var parsedBody struct {
-			ID string `json:"id"`
+			ID      string `json:"id"`
+			Name    string `json:"display_name"`
+			Country string `json:"country"`
 		}
-		err = json.Unmarshal(body, &parsedBody)
-		if err != nil {
-			common.Fail(w, fmt.Errorf("Could not parse Spotify profile response: %s", err), http.StatusInternalServerError)
+		if authenticateOnSpotify {
+			res, err := common.SpotifyClient.Do(r.Context(), func() (*http.Request, error) {
+				req, err := http.NewRequest("GET", "https://api.spotify.com/v1/me", nil)
+				if err != nil {
+					return nil, fmt.Errorf("Could not build Spotify profile request: %s", err)
+				}
+				req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", spotifyToken))
+				return req, nil
+			})
+			if err != nil {
+				common.Fail(w, fmt.Errorf("Could not make Spotify profile request: %s", err), http.StatusInternalServerError)
+				return
+			}
+			defer res.Body.Close()
+			if res.StatusCode != http.StatusOK {
+				common.Fail(w, fmt.Errorf("Spotify profile request responded with %d", res.StatusCode), http.StatusForbidden)
+				return
+			}
+			body, err := ioutil.ReadAll(res.Body)
+			if err != nil {
+				common.Fail(w, fmt.Errorf("Could not read Spotify profile response: %s", err), http.StatusInternalServerError)
+				return
+			}
+			err = json.Unmarshal(body, &parsedBody)
+			if err != nil {
+				common.Fail(w, fmt.Errorf("Could not parse Spotify profile response: %s", err), http.StatusInternalServerError)
+				return
+			}
+			_, err = redisConn.Do("HMSET", getSessionKey(spotifyToken),
+				"id", parsedBody.ID,
+				"name", parsedBody.Name,
```

Comment:
We shouldn't store `name`, we should have that in the DB once the user finalizes registration.